src/psyclone/psyir/transformations/psy_data_trans.py

```diff
@@ -0,0 +1,192 @@
+#
+# Copyright (c) 2017-2019, Science and Technology Facilities Council.
+# All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are met:
+#
+# * Redistributions of source code must retain the above copyright notice, this
+#   list of conditions and the following disclaimer.
+#
+# * Redistributions in binary form must reproduce the above copyright notice,
+#   this list of conditions and the following disclaimer in the documentation
+#   and/or other materials provided with the distribution.
+#
+# * Neither the name of the copyright holder nor the names of its
+#   contributors may be used to endorse or promote products derived from
+#   this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+# COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Authors J. Henrichs, Bureau of Meteorology
+
+'''Contains the PSyData transformation.
+'''
+
+from psyclone.psyGen import Node, Schedule
+from psyclone.psyir.transformations.region_trans import RegionTrans
+from psyclone.psyir.transformations.transformation_error \
+    import TransformationError
+from psyclone.undoredo import Memento
+
+
+class PSyDataTrans(RegionTrans):
+    ''' Create a PSyData region around a list of statements. For
+    example:
+
+    >>> from psyclone.parse.algorithm import parse
+    >>> from psyclone.parse.utils import ParseError
+    >>> from psyclone.psyGen import PSyFactory, GenerationError
+    >>> api = "gocean1.0"
+    >>> filename = "nemolite2d_alg.f90"
+    >>> ast, invokeInfo = parse(filename, api=api, invoke_name="invoke")
+    >>> psy = PSyFactory(api).create(invokeInfo)
+    >>>
+    >>> from psyclone.psyGen import TransInfo
+    >>> t = TransInfo()
+    >>> data_trans = t.get_trans_name('PSyDataTrans')
+    >>>
+    >>> schedule = psy.invokes.get('invoke_0').schedule
+    >>> schedule.view()
+    >>>
+    >>> # Enclose all children within a single profile region
+    >>> newschedule, _ = data_trans.apply(schedule.children)
+    >>> newschedule.view()
+
+    '''
+    # Unlike other transformations we can be fairly relaxed about the nodes
```

Comment:
Is this entirely true? Will the code dependency analyser work with e.g. variables passed to routines as arguments?

Reply:
Yes and no ... it will work (since any node in the PSyIR will have the ``reference_accesses`` method, but it doesn't make any sense obviously :) But afaik there are other transforms not restriction their usage either, with the same issue.

I would actually recommend that we add a class 'Statement', base on Node, and then make all Statements based on Statement instead of Node. This way we could easily restrict the node list to be for Statements only. But that would be a different issue.